### PR TITLE
Do not retry when projectId & secret don't match

### PR DESCRIPTION
**Pull request by passsy**
While testing the `example` I noticed that sending a request is retried quite often although it is clear that it will never succeed with wrong credentials.

Screen-Shot-2020-11-29-18-47-54 00

A `401` from the server now throws `UnauthenticatedWiredashApiException` which can be caught by the `RetryingFeedbackSubmitter`, preventing unnecessary retries

Refactored the networking part by joining the `NetworkManager` & `ApiClient` to become the new `WiredashApi`.

### CODE REVIEW

File: lib/src/feedback/data/retrying_feedback_submitter.dart [hunk 1]
```diff
@@ -3,7 +3,8 @@ import 'dart:math' as math;
 import 'dart:typed_data';
 
 import 'package:file/file.dart';
-import 'package:wiredash/src/common/network/network_manager.dart';
+import 'package:flutter/cupertino.dart';
```
Comment (flschweiger) [inline lib/src/feedback/data/retrying_feedback_submitter.dart:6]: I guess we don't need anything specific from `cupertino.dart`?

**Review by flschweiger — APPROVED** [review]
Nice work, looks good 👌

File: lib/src/common/network/wiredash_api.dart [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    request.headers['project'] = 'Project $projectId';
+    request.headers['authorization'] = 'Secret $secret';
+
+    final streamedResponse = await httpClient.send(request);
+    return Response.fromStream(streamedResponse);
+  }
+}
+
+/// Generic error from the Wiredash API
+class WiredashApiException implements Exception {
```
Comment (atoka93) [inline lib/src/common/network/wiredash_api.dart:69]: What about having different files for the `Exception`s? 🤔

File: lib/src/common/network/wiredash_api.dart [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+class WiredashApi {
+  WiredashApi({
+    @required this.httpClient,
+    @required this.projectId,
+    @required this.secret,
+  });
+
+  final Client httpClient;
+  final String projectId;
+  final String secret;
```
Comment (atoka93) [inline lib/src/common/network/wiredash_api.dart:17]: These could be `_private`.

File: lib/src/feedback/data/retrying_feedback_submitter.dart [hunk 4]
```diff
@@ -102,16 +106,55 @@ class RetryingFeedbackSubmitter {
...
+            ],
+          );
+          FlutterError.onError?.call(errorDetails);
           break;
         }
-      }
 
-      await Future.delayed(_exponentialBackoff(attempt));
+        // Report error and retry with exponential backoff
+        final errorDetails = FlutterErrorDetails(
```
Comment (atoka93) [inline lib/src/feedback/data/retrying_feedback_submitter.dart:145]: There could be a method for creating the errors to get rid of the duplicated code.

File: lib/src/common/network/wiredash_api.dart (hunk 2 above)
Comment (passsy) [reply lib/src/common/network/wiredash_api.dart:71]: I'd like to leave them here for now. It's the only file where they are thrown

Comment (atoka93) [reply lib/src/common/network/wiredash_api.dart:74]: 👌

File: lib/src/feedback/data/retrying_feedback_submitter.dart [hunk 5]
```diff
@@ -102,16 +107,39 @@ class RetryingFeedbackSubmitter {
...
+        await _api.sendFeedback(
+            feedback: item.feedbackItem, screenshot: screenshot);
         await _pendingFeedbackItemStorage.clearPendingItem(item);
         break;
-      } catch (_) {
+      } on UnauthenticatedWiredashApiException catch (e, stack) {
+        // Project configuration is off, retry at next app start
+        FlutterError.onError?.call(FlutterErrorDetails(
+            exception: e, stack: stack, library: 'wiredash'));
+        final errorDetails = FlutterErrorDetails(
```
Comment (atoka93) [inline lib/src/feedback/data/retrying_feedback_submitter.dart:118]: Can this be replaced by `reportWiredashError ` as well?

Comment (passsy) [reply lib/src/feedback/data/retrying_feedback_submitter.dart]: How could I miss that?

**Pull request merged**